src/Cli/dotnet/PerformanceLogEventSource.cs

```diff
@@ -0,0 +1,341 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.Tools.Common;
+using RuntimeEnvironment = Microsoft.DotNet.Cli.Utils.RuntimeEnvironment;
+
+namespace Microsoft.DotNet.Cli
+{
+    [EventSource(Name = "Microsoft-Dotnet-CLI-Performance")]
+    internal sealed class PerformanceLogEventSource : EventSource
```

Comment:
I tried to write a unit test to test the logging. But it won't work. Could you help me do that? https://github.com/dotnet/sdk/commit/5a34fc1c585d02558aa2137fefb14bc3efd2b72d

Reply:
If we want to enable it by default. I hope we have a test for every single line of logging like what I did for [this](
https://github.com/dotnet/sdk/blob/b1223209644d900702287faea8e9b71f95ec49f8/src/Tests/dotnet.Tests/TelemetryCommandTest.cs). Due to the problem of if it does not work, we will not notice until 1 year later.

Reply:
Sure.  The key here is that the `EventListener` needs to enable the `EventSource`.  This is done my implementing `OnEventSourceCreated`:

```
        protected override void OnEventSourceCreated(EventSource eventSource)
        {
            if (eventSource.Name.Equals("Microsoft-Dotnet-CLI-Performance"))
            {
                EnableEvents(eventSource, EventLevel.Verbose, EventKeywords.All);
            }
        }
```

---
In reply to: [505747573](https://github.com/wli3/sdk/pull/1#discussion_r505747573) [](ancestors = 505747573)

Reply:
Can you tell me a bit more about this?  Are you concerned that some events may not get emitted properly?  One thing that we could do (we do something like this in PerfView/TraceEvent) is do a run, capture the output file, and compare it to a baseline file.  We ignore timestamps and things that might reasonably change, but if the set of events emitted for a given run changes, then we fail the test.  This forces developers to investigate, and if necessary update the baseline.  Does that sound like it would address your concern?

---
In reply to: [505754206](https://github.com/wli3/sdk/pull/1#discussion_r505754206) [](ancestors = 505754206)

Reply:
Thanks, now the test works.

> Can you tell me a bit more about this? Are you concerned that some events may not get emitted properly?

For code quality. Like any code, it should have unit test coverage against code rot. And it need even higher code quality due to the very slow feed back loop. The log is doing a lot of IO and calling "exotic" APIs that is more complex than many main functions. Plus, we have plan to add a lot in the future, so the churn is also high. Although, at the same time, we could lower the expectation since this is for internal consumption. Say after 3 years, 1 of 20 log is missing due to refactoring of production code. And "get memory" no longer work due to a mistake from people who are not familiar with the code. Is this still ok? Considering we can still get a lot of information from the remining logs. @davidfowl @DamianEdwards @marcpopMSFT since you are on the consuming side. Require line by line code coverage is a high bar. But if it is for internal use, and we accept that, we don't need to hold it to production quality.

> capture the output file, and compare it to a baseline file.

Were you be able to run it in CI every time? I think it will not be as nice as unit tests. The test could easily be flaky due to production change, say we move some init code to a different thread. The log would show up in a different file. And it is hard to investigate why. Also, CLI code is not "deep" but "wide", meaning you need multiple run with different commands to exercise all the code.

Reply:
Definitely agree that if we're going to enable this by default, that there should be more test coverage.  Given that we're trying to start opt-in so that we can iterate quickly and learn, I'm inclined not to worry about this too much.  It sounds like from conversations with @marcpopMSFT that you and he are OK with this.  Does that sound right?

On the strategy for testing the events, this is something that we do for TraceEvent today, and yes, these tests run in the CI.  Basically, the strategy would be to write tests that run various CLI commands, and generate logs.  Then, scrutinize the logs to make sure that they contain everything we want.  Once this is done, we save them as baseline logs, and use them to compare against subsequent runs (this is done via a test harness - it's a unit test strategy).  This ensures that the set of events we expect is there.  We can also layer data validation on top of this.  Does that make sense?  Happy to explain further.

Reply:
> the strategy would be to write tests that run various CLI commands, and generate logs. Then, scrutinize the logs to make sure that they contain everything we want. Once this is done, we save them as baseline logs, and use them to compare against subsequent runs 

Previous comment on this:
Were you be able to run it in CI every time? I think it will not be as nice as unit tests. The test could easily be flaky due to production change, say we move some init code to a different thread. The log would show up in a different file. And it is hard to investigate why. Also, CLI code is not "deep" but "wide", meaning you need multiple run with different commands to exercise all the code.

I don't have good impression on this kind of functional approval tests. I've worked on similar test strategy (although for web API) that end up spend more time to update the tests and figure out why it breaks than the feature itself.